### ISSUE TITLE
Replacing uses of single-request {httr} functions with RETRY()

### ISSUE THREAD

**Issue by jameslamb** [chircollab/chircollab20#1]
**TL;DR**

There are a ton HTTP clients on CRAN, and many of them make single pass-or-fail HTTP requests. If they added some retry logic (with `httr::RETRY()`), they would be more resilient to transient issues like brief network disruptions or service downtime.

**Details**

`{httr}` is a super-popular package for making HTTP requests in R and handling their response. It has a LOT of [direct reverse imports](https://cran.r-project.org/web/packages/httr/index.html), and impacts a lot of other projects indirectly.

![image](https://user-images.githubusercontent.com/7608904/74580375-19b9f880-4f69-11ea-8e3c-e68d2be67866.png)

I have seen many examples of packages where people use functions like `httr::GET()`, `httr::POST()`, `httr::VERB()`, etc. to make HTTP requests. These functions attempt to make a single request and raise an error if anything goes wrong.

I believe (though I don't know for sure) that this all-or-nothing approach is not something package authors have carefully and intentionally chosen, and changing those calls to `httr::RETRY()` would make those packages more resilient to transient problems like brief network outages or periods where the service(s) it hits are overwhelmed. In my experience, using retry logic can improve the user experience with HTTP clients.

I propose a group spend some time improving the ecosystem of HTTP clients in R by repeating these steps:

1. choose a package from the `{httr}`  [reverse imports](https://cran.r-project.org/web/packages/httr/index.html)
2. go to the package's source code on GitHub (if it exists)
3. find all references to `httr::GET()`, `httr::POST()`, `httr::PUT()`, `httr::VERB()`, `httr::DELETE()`, `httr::HEAD()`, and `httr::PATCH()` and replace them with the appropriate `httr::RETRY()`
4. create a pull request with the change, including some text in the summary explaining to the package author how this change will make their code more reliable (see https://github.com/ropensci/eia/pull/5 for an example)

I'd be happy to start out with a short introduction to HTTP and what `httr::RETRY()` actually does.

**Comment by emilyriederer** [comment]
Thanks for posting, @jameslamb ! I love this idea of taking the time to shore up bits and pieces of many different packages and spreading best practices. 

Another good outcome might be seeing if RStudio would take a PR to `{httr}` 's 'Best Practices for API packages' vignette to share this advice more broadly

**Comment by jdblischak** [comment]
My workflowr package is one of those reverse dependencies, and I super welcome a PR to make the GitHub API calls more robust.

The function [`create_gh_repo()`](https://github.com/jdblischak/workflowr/blob/master/R/wflow_use_github.R#L372) makes 3 `GET` requests and a `POST` request.

**Comment by jameslamb** [comment]
I wrote up a 101-level "wtf is HTTP" last night: https://github.com/jameslamb/talks/tree/master/chi-r-collab-httr. Decided to put it in that repo because I might turn it into a lightning talk in the future.

For those who want to do the "go make PRs to existing packages" part of this project, here are some examples:

* https://github.com/ropensci/plotly/pull/1656
* https://github.com/cloudyr/aws.alexa/pull/15
* https://github.com/ropensci/eia/pull/5
* https://github.com/r-transit/tidytransit/pull/131
* https://github.com/willdebras/ctar/pull/9
* https://github.com/uptake/uptasticsearch/pull/172
* https://github.com/dmi3kno/polite/commit/7d3428081eb359114305d67f8bbbf7a11b079dbd

The basic process is:

1. Go to [httr's CRAN page](https://cran.r-project.org/web/packages/httr/index.html)
2. Look at the packages in `Reverse Depends` and `Reverse Imports`. Choose one that sounds interesting and click on it.
3. Hopefully the package has a `URL:` section with a GitHub link. Click it.
4. From that repository, click `fork` to fork it. ([GitHub docs](https://help.github.com/en/github/getting-started-with-github/fork-a-repo))
5. Clone the repository to your machine.([GitHub docs](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/cloning-a-repository))
6. Look for uses of `httr` single-shot functions. 
    - These are:
         - `DELETE()`
         - `GET()`
         - `HEAD()`
         - `POST()`
         - `PATCH()`
         - `PUT()`
         - `VERB()`
    - You can find these with commands like `git grep 'POST('` or using [Find in Files in RStudio](https://support.rstudio.com/hc/en-us/articles/200710523-Navigating-Code)
7. Replace those calls with `httr::RETRY()`. See the pull requests linked above for some examples.
8. Commit your changes, push them to your fork, and create a pull request ([GitHub docs](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request))
    - When you open the PR, be sure to add a link to this issue in the description. That way GitHub will do the hard work of tracking for us, like this!

![image](https://user-images.githubusercontent.com/7608904/79641885-e3167f00-815f-11ea-9b61-0c42dca6aab3.png)

**Comment by jameslamb** [comment]
Some resources  shared during our first  chat:

*  https://http.cat
* http://httpbin.org (try http://httpbin.org/status/418 😀 )
* Mozilla has a page for every status code, for example:  https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/418

**Comment by jameslamb** [comment]
Cleaning up my "created by me" issue list today, I think this served it's purpose and can be closed.

Thanks for all your help everyone, this was fun!!!
